File: Service/Translators/UnitGroupTypeTranslator.vb
Imports BO
Imports DAL

Public Class UnitGroupTypeTranslator
    Public Shared Function TranslateEntityToBO(_entity As UnitGroupTypes, bo As UnitGroupTypeBO) As ErrorCode
        If _entity Is Nothing Then Return ErrorCode.EntityNull
        If bo Is Nothing Then Return ErrorCode.BoNull
        bo.Id = _entity.ID
        bo.Name = _entity.Name
        If (_entity.UnitTypes IsNot Nothing) Then
            For Each Type In _entity.UnitTypes
                Dim UnitType As New UnitTypeBO
                Dim err = UnitTypeTranslator.TranslateEntityToBO(Type, UnitType)
                If (err <> ErrorCode.Success) Then Return err
                bo.UnitTypes.Add(UnitType)
            Next

        End If
        Return ErrorCode.Success
    End Function

    Friend Shared Function TranslateBOToEntity(_entity As UnitGroupTypes, bo As UnitGroupTypeBO) As ErrorCode
        If _entity Is Nothing Then Return ErrorCode.EntityNull
        If bo Is Nothing Then Return ErrorCode.BoNull
        _entity.Name = bo.Name
        Dim err = HandleTypes(_entity, bo.UnitTypes)
        If (err <> ErrorCode.Success) Then Return err
        Return ErrorCode.Success
    End Function
    Private Shared Function HandleTypes(_entity As UnitGroupTypes, types As List(Of UnitTypeBO)) As ErrorCode
        If (types.Count = 0) Then Return ErrorCode.Success
        For Each x In types
            If (x.Id = 0) Then
                'insert
                Dim type As New UnitTypes
                Dim err = UnitTypeTranslator.TranslateBOToEntity(type, x)
                If (err <> ErrorCode.Success) Then Return err
                _entity.UnitTypes.Add(type)
            Else
                'update
                Dim type = _entity.UnitTypes.FirstOrDefault(Function(f) f.ID = x.Id)
                If (type IsNot Nothing) Then
                    Dim err = UnitTypeTranslator.TranslateBOToEntity(type, x)
                    If (err <> ErrorCode.Success) Then Return err
                End If
            End If
        Next
        'delete
        Dim delList As New List(Of UnitTypes)
        For Each x In _entity.UnitTypes
            If (Not types.Any(Function(f) f.Id = x.ID)) Then
                delList.Add(x)
            End If
        Next
        For Each x In delList
            _entity.UnitTypes.Remove(x)
        Next
        Return ErrorCode.Success
    End Function
End Class

File: ServiceCore/Translators/UnitGroupTypeTranslator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using BOCore;
using DALCore.Models;
using DALCore;

namespace ServiceCore.Translators
{
    public class UnitGroupTypeTranslator
    {
        public static ErrorCode TranslateEntityToBO(UnitGroupTypes _entity, UnitGroupTypeBO bo)
        {
            if (_entity == null)
                return ErrorCode.EntityNull;
            if (bo == null)
                return ErrorCode.BoNull;
            bo.Id = _entity.Id;
            bo.Name = _entity.Name;
            if ((_entity.UnitTypes != null))
            {
                foreach (var Type in _entity.UnitTypes)
                {
                    UnitTypeBO UnitType = new UnitTypeBO();
                    var err = UnitTypeTranslator.TranslateEntityToBO(Type, UnitType);
                    if ((err != ErrorCode.Success))
                        return err;
                    bo.UnitTypes.Add(UnitType);
                }
            }
            return ErrorCode.Success;
        }

        internal static ErrorCode TranslateBOToEntity(UnitGroupTypes _entity, UnitGroupTypeBO bo)
        {
            if (_entity == null)
                return ErrorCode.EntityNull;
            if (bo == null)
                return ErrorCode.BoNull;
            _entity.Name = bo.Name;
            var err = HandleTypes(_entity, bo.UnitTypes);
            if ((err != ErrorCode.Success))
                return err;
            return ErrorCode.Success;
        }
        private static ErrorCode HandleTypes(UnitGroupTypes _entity, List<UnitTypeBO> types)
        {
            if ((types.Count == 0))
                return ErrorCode.Success;
            foreach (var x in types)
            {
                if ((x.Id == 0))
                {
                    // insert
                    UnitTypes type = new UnitTypes();
                    var err = UnitTypeTranslator.TranslateBOToEntity(type, x);
                    if ((err != ErrorCode.Success))
                        return err;
                    _entity.UnitTypes.Add(type);
                }
                else
                {
                    // update
                    var type = _entity.UnitTypes.FirstOrDefault(f => f.Id == x.Id);
                    if ((type != null))
                    {
                        var err = UnitTypeTranslator.TranslateBOToEntity(type, x);
                        if ((err != ErrorCode.Success))
                            return err;
                    }
                }
            }
            // delete
            List<UnitTypes> delList = new List<UnitTypes>();
            foreach (var x in _entity.UnitTypes)
            {
                if ((!types.Any(f => f.Id == x.Id)))
                    delList.Add(x);
            }
            foreach (var x in delList)
                _entity.UnitTypes.Remove(x);
            return ErrorCode.Success;
        }
    }
}
